Storage/StorageManager.php

```diff
@@ -161,68 +118,177 @@ public function storeUploadedFile($type, UploadedFile $uploadedFile, $newFilenam
      */
     public function storeFileFromLocalPath($type, $path, $newFilename = null, $unlinkAfterStore = false)
     {
-        if (!file_exists($path)) {
-            throw new \InvalidArgumentException(sprintf("Cannot store file from non-existent path at '%s'", $path));
+        $options = [];
+        if ($newFilename !== null) {
+            $options['newFilename'] = $newFilename;
         }
-        $this->validateFileStorageType($type);
-
-        $storage = $this->getStorageFromWritePolicy($type);
+        if ($unlinkAfterStore !== false) {
+            $options['unlinkAfterStore'] = $unlinkAfterStore;
+        }
+        return $this->store($type, $path, $options);
+    }
 
-        if (null === $newFilename) {
-            $newFilename = uniqid('gen' . substr(hash('sha256', $path), 0, 4));
+    /**
+     * Takes a content and store it in permanent storage
+     * Content might be:
+     * UploadedFile
+     * string - path to local file
+     * StreamInterface - stream containing file contents
```

Comment:
can we also support resources?

Reply:
Yes, we can, but it will require a completely new piece of code. resource can't be passed to file->setContent directly. So, in my opinion, this implementation can go into another PR.

Reply:
Right, but $file->setContent isn't the only way to write.  We can use the ioStream implementation for resources I think.

Reply:
You are right, but I insist that it should go in another PR, because with pure resource it is not clear how we could obtain md5 of the file. While in current implementation we rely on assumption that Stream::__toString will return whole content, with 'resource' type we would need to do something like 'fseek($resource, 0)' after uploading, then save to temporary file, and then generate md5 hash. Moreover with such resource as 'php://input' this would not work, because while it says that this stream is seekable, it appears that it is not. You should open new resource if you want to reread contents.

The other way to implement this is to use CachedStream from guzzlehttp (or similar). It will allow us to cache contents in memory, and then use them to generate the hash. We could even write contents to a temporary file to reduce memory usage.

In any case this would require significant amount of code and testing.

BTW in our case, we use CachedStream to pass it to storage bundle, precisely because of issues with hash generation.